### PR TITLE
Pull request 6

### PR DESCRIPTION
**Pull request by vitsalis**

### CODE REVIEW

File: models/user.php [hunk 1]
```diff
@@ -45,7 +45,7 @@ protected function validate() {
             if ( empty( $this->username ) ) {
                 throw new ModelValidationException( 'username_empty' );
             }
-            if ( strpos( $this->username, ' ' ) || preg_match( '#[^a-zA-Z0-9]#', $this->username ) ) {
+            if ( strpos( $this->username, ' ' ) || preg_match( '#[^a-zA-Z0-9._]#', $this->username ) ) {
```
Comment (dionyziz) [inline models/user.php]: strpos must be checked against boolean false using ===, not like this; suggest moving space into regexp

File: database/db_schema.sql [hunk 2]
```diff
@@ -31,7 +31,46 @@ CREATE TABLE IF NOT EXISTS `countries` (
...
+
+-- --------------------------------------------------------
+
+--
+-- Table structure for table `games`
+--
+
+CREATE TABLE IF NOT EXISTS `games` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `created` date NOT NULL,
```
Comment (dionyziz) [inline database/db_schema.sql]: DATETIME

File: database/db_schema.sql [hunk 3]
```diff
@@ -42,9 +81,26 @@ CREATE TABLE IF NOT EXISTS `countries` (
...
+
+--
+-- Table structure for table `roundcreatures`
+--
+
+CREATE TABLE IF NOT EXISTS `roundcreatures` (
+  `roundid` int(11) NOT NULL,
+  `gameid` int(11) NOT NULL,
+  `creatureid` int(11) NOT NULL,
+  `desire` enum('NORTH','WEST','EAST','SOUTH','NONE') COLLATE utf8_unicode_ci NOT NULL,
```
Comment (dionyziz) [inline database/db_schema.sql]: Would prefer to order this as per CSS's order: NONE, NORTH, EAST, SOUTH, WEST

File: database/db_schema.sql (hunk 2 above)
Comment (vitsalis) [inline database/db_schema.sql]: You said it's date because we have a game per day

File: views/user/view.php [hunk 4]
```diff
@@ -6,6 +6,10 @@
             echo $user->image->target_path;
         ?>" alt="Profile Picture" width="100" height="100" /></p>
 
+<p><img src="<?php
+            echo $user->country->flag;
+        ?>" alt="Country flag" width="100" height="100" /></p>
```
Comment (dionyziz) [inline views/user/view.php]: Use country name in alt.

File: database/db_schema.sql (hunk 2 above)
Comment (dionyziz) [inline database/db_schema.sql]: We might have more games in a single day later. Furthermore, it's good for logistic purposes to also record the time of the game, even if it happens once a day. We can display only the day to the users if we wish.

**Comment by dionyziz** [comment]
Please undo the unnecessary chmod changes.

File: controllers/dashboard.php [hunk 5]
```diff
@@ -1,9 +1,12 @@
 <?php
-    class DashboardController {
-        public static function view() {
+    class DashboardController extends ControllerBase {
+        public function view() {
             if ( isset( $_SESSION[ 'user' ] ) ) {
                 $username = $_SESSION[ 'user' ][ 'username' ];
             }
+            include_once 'models/formtoken.php';
+            $token = FormToken::create();
```
Comment (dionyziz) [inline controllers/dashboard.php]: DRY. This should be the responsibility of the Form class.

File: helpers/html.php [hunk 6]
```diff
@@ -1,4 +1,96 @@
...
+        private $method;
+        public $id;
+        public $token;
+
+        public function __construct( $resource = '', $method = '' ) {
+            $this->resource = $resource;
+            $this->method = $method;
+        }
+
+        public static function produceError( $error_msg ) {
```
Comment (dionyziz) [inline helpers/html.php]: Why is this "produce" and the one below "create"?

File: helpers/html.php [hunk 7]
```diff
@@ -1,4 +1,96 @@
...
+            ?><p class="error"><?php
+                    echo $error_msg;
```
Comment (dionyziz) [inline helpers/html.php]: indentation

File: helpers/html.php [hunk 8]
```diff
@@ -1,4 +1,96 @@
...
+            ?></p><?php
+        }
+
+        public static function createInput( $type = 'text', $name = '', $id = '', $value = '' ) {
+            ?><p><input type="<?php
+                echo $type;
```
Comment (dionyziz) [inline helpers/html.php:62]: whitelist types

File: helpers/html.php [hunk 9]
```diff
@@ -1,4 +1,96 @@
...
+                    ?>" <?php
+                }
+                if ( !empty( $id ) ) {
+                    ?>id="<?php
+                        echo $id;
+                    ?>" <?php
+                }
+                if ( !empty( $value ) ) {
+                    ?>value="<?php
+                        echo $value;
```
Comment (dionyziz) [inline helpers/html.php]: XSS

File: helpers/html.php [hunk 10]
```diff
@@ -1,4 +1,96 @@
...
+                    ?>id="<?php
+                        echo $id;
+                    ?>" <?php
+                }
+            ?>><?php
+            foreach ( $option_array as $option ) {
+                ?><option <?php
+                    if ( isset( $option[ 'value' ] ) ) {
+                        ?>value="<?php
+                            echo $option[ 'value' ];
```
Comment (dionyziz) [inline helpers/html.php]: XSS

File: helpers/html.php [hunk 11]
```diff
@@ -1,4 +1,96 @@
...
+                ?></option><?php
+            }
+            ?></select></p><?php
+        }
+       
+        public static function createLabel( $for, $text ) {
+            ?><label for="<?php
+                echo $for;
+            ?>"><?php
+                echo $text;
```
Comment (dionyziz) [inline helpers/html.php]: XSS

File: helpers/html.php [hunk 12]
```diff
@@ -1,4 +1,96 @@
...
+                if ( isset( $this->id ) ) {
+                    ?>id="<?php
+                        echo $this->id;
+                    ?>" <?php
+                }
+                ?>action="index.php?resource=<?php
+                    echo $this->resource;
+                ?>&amp;method=<?php
+                    echo $this->method;
+                ?>" method="POST"><?php
```
Comment (dionyziz) [inline helpers/html.php]: - Lower-case "post".
- Some forms may need to be "get".
- This doesn't work for file uploads; detect if there are file inputs and send correct enctype

File: helpers/html.php [hunk 13]
```diff
@@ -1,4 +1,96 @@
 <?php
+    class Form {
+        private $resource;
+        private $method;
+        public $id;
+        public $token;
```
Comment (dionyziz) [inline helpers/html.php]: protect \$token and own it; the caller should not concern themselves with this

File: helpers/html.php [hunk 14]
```diff
@@ -1,4 +1,96 @@
 <?php
+    class Form {
+        private $resource;
+        private $method;
```
Comment (dionyziz) [inline helpers/html.php]: make these protected for extensibility

File: views/session/create.php [hunk 15]
```diff
@@ -2,35 +2,33 @@
...
-<div id="login"><?php
+<div id="login"><?php 
+    $form = new Form( 'session', 'create' );  
+    $form->id = 'login-form';
+    $form->token = $token;
+    $form->output( function() use ( $username_empty, $password_empty,
+            $username_wrong, $password_wrong ) {
         if ( isset( $username_empty ) ) {
-            ?><p class="error">Please type a username.</p><?php
+            Form::produceError( "Please type a username." );
```
Comment (dionyziz) [inline views/session/create.php]: Make these non-static method calls. Form needs to maintain state.

File: views/session/create.php [hunk 16]
```diff
@@ -2,35 +2,33 @@
...
-        ?>
-        <p><input type="text" name="username" id="username" /></p>
-        <label for="password">Password</label>
-        <?php
-            if ( isset( $password_wrong ) ) {
-                ?><p class="error">Password is incorrect</p><?php
-            }
-        ?>
-        <p><input type="password" name="password" id="password" /></p>
-        <p><a href="">Forgot password?</a></p>
```
Comment (dionyziz) [inline views/session/create.php:29]: Where did this go?

File: views/user/create.php [hunk 17]
```diff
@@ -20,128 +20,95 @@
...
-                    ?></option><?php
-                }
-            ?>
-        </select> 
-    </p>
-    <p><input type="submit" value="Register" /></p>
-</form>
+            $email_value = "";
+        }
+        Form::createInput( 'text', 'email', 'email', htmlspecialchars( $email_value ) );
```
Comment (dionyziz) [inline views/user/create.php]: Move htmlspecialchars to form; pass actual value here.

File: views/user/create.php [hunk 18]
```diff
@@ -20,128 +20,95 @@
...
+        Form::createLabel( 'dob', 'Date of birth' );
+        $days_select_array = array();
+        $days_select_array[] = array( 'content' => 'Select Day' );
```
Comment (dionyziz) [inline views/user/create.php]: Combine these two lines.

File: views/user/create.php [hunk 19]
```diff
@@ -20,128 +20,95 @@
...
+        for ( $i = 1; $i <= 31; ++$i ) {
+            $days_select_array[] = array( 'value' => $i, 'content' => $i );
+        }
+        Form::createSelect( 'day', 'dob', $days_select_array );
+        $months_select_array = array();
+        $months_select_array[] = array( 'content' => 'Select Month' );
```
Comment (dionyziz) [inline views/user/create.php]: Combine

File: views/user/create.php [hunk 20]
```diff
@@ -20,128 +20,95 @@
...
+        for ( $i = 1; $i <= 12; ++$i ) {
+            $months_select_array[] = array( 
+                'value' => $i, 
+                'content' => date( 'M', mktime( 0, 0, 0, $i, 1, 2000 ) ) 
+            );
+        }
+        Form::createSelect( 'month', '', $months_select_array );
+        $years_select_array = array();
+        $years_select_array[] = array( 'content' => 'Select Year' );
```
Comment (dionyziz) [inline views/user/create.php]: Combine

File: views/user/create.php [hunk 21]
```diff
@@ -20,128 +20,95 @@
...
+        $current_year = date( 'Y' );
+        for ( $i = $current_year - 8; $i >= $current_year - 100; --$i ) {
```
Comment (dionyziz) [inline views/user/create.php]: min age / max age should be configurable constants

File: views/user/create.php [hunk 22]
```diff
@@ -20,128 +20,95 @@
...
+            $years_select_array[] = array( 'value' => $i, 'content' => $i );
+        }
+        Form::createSelect( 'year', '', $years_select_array );
+        $countries_select_array = array();
+        $countries_select_array[] = array( 'content' => 'Select Country' );
```
Comment (dionyziz) [inline views/user/create.php]: combine

File: views/user/create.php [hunk 23]
```diff
@@ -20,128 +20,95 @@
...
+        foreach ( $countries as $key => $country ) {
+            $countries_select_array[] = array( 'value' => $key + 1, 'content' => $country[ 'name' ] );
```
Comment (dionyziz) [inline views/user/create.php:104]: Why +1?

File: views/user/update.php [hunk 24]
```diff
@@ -1,72 +1,61 @@
...
+        Form::createLabel( 'image', 'Upload an avatar' );
         if ( isset( $image_invalid ) ) {
-            ?><p class="error">This isn't an image</p><?php
+            Form::produceError( "This isn't an image" );
         }
-    ?>
-    <p><input type="file" name="image" id="image" /></p>
-    <input type="submit" value="Upload" />
-</form>
+        Form::createInput( 'file', 'image', 'image' );
```
Comment (dionyziz) [inline views/user/update.php]: Did you test this?

File: views/session/create.php (hunk 15 above)
Comment (vitsalis) [inline views/session/create.php]: how am I going to do that?

File: views/user/create.php (hunk 23 above)
Comment (vitsalis) [inline views/user/create.php:104]: because at database the countries id starts from 1 but here they keys start from 0

File: helpers/html.php [hunk 25]
```diff
@@ -1,4 +1,147 @@
...
+                'reset',
+                'search',
+                'submit',
+                'tel',
+                'text',
+                'time',
+                'url',
+                'week'
+            );
+            foreach ( $valid_types as $valid_type ) {
```
Comment (dionyziz) [inline helpers/html.php]: array_flip() and isset()

File: views/user/update.php [hunk 26]
```diff
@@ -1,72 +1,59 @@
...
-</form>
+        Form::createSelect( 'countryid', '', $countries_select_array );
+        Form::createInput( 'submit', '', '', 'Save settings' );
+    } );
 
-<form action="index.php?resource=image&amp;method=create" method="post" enctype="multipart/form-data"> 
-    <label for="image">Upload an avatar</label>
-    <?php
+    $form = new Form( 'image', 'create' );
+    $form->hasFile = true;
```
Comment (dionyziz) [inline views/user/update.php]: Don't set this here. Let the form take care of it when you createInput( 'file' ... );

File: helpers/html.php [hunk 27]
```diff
@@ -1,4 +1,147 @@
 <?php
+    include_once 'models/formtoken.php';
+    class Form {
+        protected $resource;
+        protected $method;
+        public $id;
+        public $formMethod;
+        public $hasFile = false;
```
Comment (dionyziz) [inline helpers/html.php]: protected

File: views/session/create.php [hunk 28]
```diff
@@ -2,36 +2,36 @@
     include 'views/header.php';
 ?>
 
-<div id="login"><?php
+<div id="login"><?php 
+    $form = new Form( 'session', 'create' );  
+    $form->id = 'login-form';
+    $form->formMethod = 'post';
```
Comment (dionyziz) [inline views/session/create.php]: Let the form handle this based on method passed to constructor. For idempotent methods, use GET. For non-idempotent methods, use POST.

File: helpers/html.php [hunk 29]
```diff
@@ -1,4 +1,182 @@
...
+                'search',
+                'submit',
+                'tel',
+                'text',
+                'time',
+                'url',
+                'week'
+            );
+            $valid_types = array_flip( $valid_types );
+            if ( isset( $valid_types[ $type ] ) ) {
```
Comment (dionyziz) [inline helpers/html.php:37]: `return isset...`

**Pull request merged**